engine/execution/rpc/engine.go

```diff
@@ -793,3 +801,30 @@ func (h *handler) blockHeaderResponse(header *flow.Header) (*execution.BlockHead
 		Block: msg,
 	}, nil
 }
+
+// additional check that when there is no event in the block, double check if the execution
+// result has no events as well, otherwise return an error.
+// we check the execution result has no event by checking if each chunk's EventCollection is
+// the default hash for empty event collection.
+func (h *handler) getEventsByBlockID(blockID flow.Identifier) ([]flow.Event, error) {
+	blockEvents, err := h.events.ByBlockID(blockID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get events for block: %v", err)
+	}
+
+	if len(blockEvents) == 0 {
+		executionResult, err := h.exeResults.ByBlockID(blockID)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to get execution result for block %v: %v", blockID, err)
+		}
+
+		for _, chunk := range executionResult.Chunks {
+			if chunk.EventCollection != flow.EmptyEventCollectionID &&
+				executionResult.PreviousResultID != flow.ZeroID { // skip the root blcok
```

Comment:
Root Block uses flow.ZeroID (instead of flow.EmptyEventCollectionID) as EventCollection, so we skip the root block here.